Test/domains/urcu-bp.hpp

```diff
@@ -3,22 +3,23 @@
 #include "rcu_domain.hpp"
 
 #include <urcu-bp.h>
-#define _LGPL_SOURCE
 
-namespace std {
-    class rcu_bp: public rcu_domain {
-    public:
-        void register_thread() { rcu_register_thread(); }
-        void unregister_thread() { rcu_unregister_thread(); }
-        void read_lock() noexcept { rcu_read_lock(); }
-        void read_unlock() noexcept { rcu_read_unlock(); }
-        void synchronize() noexcept { synchronize_rcu(); }
-        void retire(rcu_head *rhp, void (*cbf)(rcu_head *rhp)) { call_rcu(rhp, cbf); }
-        void barrier() noexcept { rcu_barrier(); }
-        void quiescent_state() noexcept { rcu_quiescent_state(); }
-        void thread_offline() noexcept { rcu_thread_offline(); }
-        void thread_online() noexcept { rcu_thread_online(); }
+class rcu_domain_bp {
+public:
+    static constexpr bool register_thread_needed() { return false; }
+    void register_thread() { rcu_register_thread(); }
+    void unregister_thread() { rcu_unregister_thread(); }
+    void thread_offline() noexcept { rcu_thread_offline(); }
+    void thread_online() noexcept { rcu_thread_online(); }
 
-        static constexpr bool register_thread_needed() { return false; }
-    };
-} // namespace std
+    static constexpr bool quiescent_state_needed() { return false; }
+    void quiescent_state() noexcept { rcu_quiescent_state(); }
```

Comment:
First, apologies for the delay -- the day job grabbed me by both ankles in several directions for quite some time.  But the pending Linux-kernel RCU commits greatly increased SRCU scalability and also do a net reduction of more than 2700 lines of code, so at least something useful came out of it...  And yes, had I been thinking at all at the time, I would have simply handed the C++ RCU reins over to you.

OK...  quiescent_state().  Suppose I have some code that doesn't care what flavor of RCU it is using.  If it happens to be using rcu_domain_qsbr, then that code absolutely must periodically invoke quiescent_state().  So either the code has be changed when the underlying RCU implementation switches to or from rcu_domain_qsbr, or it has to call quiescent_state() all the time, which means (as far as I know) that all domains must supply some sort of implementation, if only an empty function.

Does that help, or am I missing the point of your question?